### PR TITLE
Piper/initial library code

### PR DESCRIPTION
**Pull request by pipermerriam**
## What was wrong?

Implements the initial library code which was extracted from the Trinity codebase....

## How was it fixed?

Mostly copy/paste but there is also new functionality (all of which is tested)

#### Cute Animal Picture

PuppyDay_AntarcticFurSealPup_ChristopherWillis

### CODE REVIEW

**Comment by pipermerriam** [comment]
@carver My plan for this is to release an `alpha` of this, migrate @jannikluhn 's discovery v5 code over to use the slightly modified new API and remove the new trio service from the trinity codebase, then start migrating some of the less boring existing services to use the `asyncio` version of the service to start battle testing it.

File: async_service/abc.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+            async with run_service(service) as manager:
+                # service runs inside context block
+                ...
+                # service cancels and stops when context exits
+            # service will have fully stopped
+
+            # 2. run the service blocking until completion
+            await Manager.run_service(service)
+
+            # 3. create manager and then run service blocking until completion
```
Comment (carver) [inline async_service/abc.py:31]: Any intended difference between 3 and 2? Or are they fully equivalent.

**Review by carver — APPROVED** [review]
I didn't really review it, it's just too big. Given that it's an alpha, I guess we'll just roll with it :man_shrugging:

File: async_service/abc.py (hunk 1 above)
Comment (pipermerriam) [reply async_service/abc.py:31]: 2 doesn't give you a handle to the manager instance where-as 3 does.

**Pull request merged**